README
$Id: README 1966 2007-01-03 04:44:01Z comdog $

You can install this using in the usual Perl fashion

	perl Makefile.PL
	make
	make test
	make install
	
The documentation is in the module file.  Once you install
the file, you can read it with perldoc.

	perldoc HTML::SimpleLinkExtor
	
If you want to read it before you install it, you can use
perldoc directly on the module file.

	perldoc lib/SimpleLinkExtor.pm
	
This module is also in CVS on SourceForge

	http://sourceforge.net/projects/brian-d-foy/

You may use HTML::SimpleLink::Extor under the same terms as Perl
itself.

Enjoy, 

brian d foy, bdfoy@cpan.org

Changes
Revision history for Perl extension HTML::SimpleLinkExtor.

1.25 - Fri Aug  5 07:56:17 2011
	* I need Test::More 0.96 for subtests and done_testing.

1.24 - Thu Aug  4 06:37:30 2011
	* Fix clear_links so it actually clears the links
	* Convert most tests to use subtest and Test::More 0.90
	* Cleanups for the move to Github

1.23 - Mon Jun  1 01:04:37 2009
	* RT #45198: close off iframe tags in example HTML since 
	HTML::Parser >= 3.56 now ignores the content of iframes
	(so it needs to know where they stop)

1.22 - Sun Sep  7 18:02:42 2008
	* Fix a test problem with 5.6.2. It's http://rt.cpan.org/Ticket/Display.html?id=38796
	in Test::Output.

1.21_02 - Mon Sep  1 23:01:25 2008
	* More fixes to parse.t for Windows. Now I know how to correctly 
	make file: URIs for Windows:
	http://blogs.msdn.com/ie/archive/2006/12/06/file-uris-in-windows.aspx

1.21_01 - Mon Sep  1 12:22:06 2008
	* Fix some cross-platform problems with parse.t

1.21 - Sun Aug 31 19:45:20 2008
	* You can now interact with the user-agent through the ua()
	method. Change the agent string or whatever you like.

1.19 - Fri Dec  7 01:32:15 2007
	* RT #31160: added a schemes() method to extract links by their
	scheme, and absolute_links() and relative_links() to extract
	by their absoluteness. :)

1.18 - Sat Nov 24 04:44:09 2007
	* Cleanups for the move from CVS to SVN
	* No need to upgrade

1.17 - Sun Mar 18 22:32:01 2007
	* Cleaning up the dist a bit
	* Now takes input from the web with parse_url()

1.16 - Tue Jan  9 22:07:34 2007
	* updated copyright and license info
	* no code changes, so no need to upgrade

1.15 - Thu Jan  4 11:37:02 2007
	* added parse_url method to fetch remote data for link extraction
	* fixed man page destination for linktractor

1.14 - Tue Jan  2 22:44:41 2007
	* Added a command line tool, linktractor, to extract the links
	from the input files, stdin, or a file of the net.
	* Updated license info. Use this under the terms of Perl itself.
	* Happy New Year!

1.13 - Sat May 27 02:21:37 2006
	* Added functions to add tags and attributes to the list that
	the module will parse. You don't have to fool with the internal
	hash or email me to make changes anymore!
	* Unless you want this feature, you don't need to upgrade.

1.12 - Wed Jan 11 00:51:33 2006
	* rt.cpan.org: Ticket #16942: A bug in my URL resolution code
	only resolved the last URL found in a tag. Now the code goes through
	all the value and resolves each that it thinks should be a URL
	* Everyone should upgrade

1.11 - Mon Aug 22 22:44:18 2005
	* added support for iframe src extraction
	* no need to upgrade unless you need this feature

1.10 - Tue Aug 16 14:42:24 2005
	* added a clear_links() method so you can start fresh if you
	want to parse multiple files with the same parser object. This
	does not affect previous behavior, so old scripts should be
	safe.

1.09 - Fri Mar 11 22:52:40 2005
	* Added POD coverage tests, fixed issues
	* No code changes: no need to upgrade

1.08 - Tue Sep  7 20:45:10 2004
	* cleaned up distribution
	* no need to upgrade

1.06 - Tue Jun 15 13:32:38 2004
	* added test for SCRIPT tag
	* noted in docs that HTML::LinkExtor handles the parsing, not me

1.05 - Tue Jun 15 13:08:22 2004
	* added support to extract references from SCRIPT tags
	* cleaned up the dist to use Test::More, and other housekeeping things

0.72 Fri Nov  9 04:07:17 EST 2001
	- fixed bug with img tags that contained a usemap attribute
	(identified by Will Crain <crainw@hotmail.com>)

0.71  Wed Jan 24 14:34:10 EST 2001
	- minor doc fixes

0.07  Fri Nov 17 19:18:41 EST 2000
	- added support to extract frame SRC attribute values.

0.01  Sun Oct  8 00:58:58 2000
	- original version; created by h2xs 1.18



lib/SimpleLinkExtor.pm
# $Id: SimpleLinkExtor.pm 2445 2007-12-07 07:15:14Z comdog $
package HTML::SimpleLinkExtor;
use strict;

use warnings;
no warnings;

use subs qw();
use vars qw($VERSION @ISA %AUTO_METHODS $AUTOLOAD $DEBUG);

use AutoLoader;
use Carp qw(carp);
use HTML::LinkExtor;
use URI;

$VERSION = 1.19;

$DEBUG   = 0;

@ISA = qw(HTML::LinkExtor);

%AUTO_METHODS = qw(
    background attribute
	href	attribute
	src		attribute

	a		tag
	area	tag
	base    tag
	body    tag
	img		tag
	frame	tag
	iframe  tag

	script	tag
	);


sub DESTROY { 1 };

sub AUTOLOAD
	{
	my $self = shift;
	my $method = $AUTOLOAD;

	$method =~ s/.*:://;
	print STDERR "AUTOLOAD: method is $method\n" if $DEBUG;

	unless( exists $AUTO_METHODS{$method} )
		{
		carp __PACKAGE__ . ": method $method unknown";
		return;
		}

	print STDERR "AUTOLOAD: calling _extract\n" if $DEBUG;
	$self->_extract( $method );
	}

sub _init_links
	{
	my $self  = shift;
	my $links = shift;
		
	do { 
		delete $self->{'_SimpleLinkExtor_links'};
		return
		} unless UNIVERSAL::isa( $links, 'ARRAY' );
	
	$self->{'_SimpleLinkExtor_links'} = $links;
	
	$self;
	}

sub _link_refs
	{
	my $self = shift;

	my @link_refs;
	# XXX: this is a bad way to do this. I should check if the
	# value is a reference. If I want to reset the links, for
	# instance, I can't just set it to [] because it then goes
	# through this branch. In _init_links I have to use a delete
	# which I really don't like. I don't have time to rewrite this
	# right now though --brian, 20050816
	if( ref $self->{'_SimpleLinkExtor_links'} )
		{
		@link_refs = @{$self->{'_SimpleLinkExtor_links'}};
		}
	else
		{
		@link_refs = $self->SUPER::links();
		$self->_init_links( \@link_refs );
		}

	# defined() so that an empty string means "do not resolve"
	unless( defined $self->{'_SimpleLinkExtor_base'} )
		{
		my $count = -1;
		my $found =  0;
		foreach my $link ( @link_refs )
			{
			$count++;
			next unless $link->[0] eq 'base' and $link->[1] eq 'href';
			$found = 1;
			$self->{'_SimpleLinkExtor_base'} = $link->[-1];
			last;
			}

		#remove the BASE HREF link - Good idea, bad idea?
		#splice @link_refs, $count, 1, () if $found;
		}

	$self->_add_base(\@link_refs) if $self->{'_SimpleLinkExtor_base'};

	print "_link_refs: there are $#link_refs + 1 links\n" if $DEBUG;
	return @link_refs;
	}

sub _extract
	{
	my $self      = shift;
	my $method    = shift;

	my $position  = $AUTO_METHODS{$method} eq 'tag' ? 0 : 1;
	print "_extract: Position is $position\n" if $DEBUG;

	my @links = map  { $$_[2] }
	            grep { $_->[$position] eq $method }
	            $self->_link_refs;

	print "_extract: There are $#links + 1 links\n" if $DEBUG;
	return @links;
	}

sub _add_base
	{
	my $self      = shift;
	my $array_ref = shift;

	my $base      = $self->{'_SimpleLinkExtor_base'};
	next unless $base;
	
	foreach my $tuple ( @$array_ref )
		{
		foreach my $index ( 1 .. $#$tuple )
			{
			next unless exists $AUTO_METHODS{ $tuple->[$index] };
			
			my $url = URI->new( $tuple->[$index + 1] );
			next unless ref $url;
			$tuple->[$index + 1] = $url->abs($base);
			}
		}
	}

=head1 NAME

HTML::SimpleLinkExtor - Extract links from HTML

=head1 SYNOPSIS

	use HTML::SimpleLinkExtor;

	my $extor = HTML::SimpleLinkExtor->new();
	$extor->parse_file($filename);
	#--or--
	$extor->parse($html);

	$extor->parse_file($other_file); # get more links

	$extor->clear_links; # reset the link list
	
	#extract all of the links
	@all_links   = $extor->links;

	#extract the img links
	@img_srcs    = $extor->img;

	#extract the frame links
	@frame_srcs  = $extor->frame;

	#extract the hrefs
	@area_hrefs  = $extor->area;
	@a_hrefs     = $extor->a;
	@base_hrefs  = $extor->base;
	@hrefs       = $extor->href;

	#extract the body background link
	@body_bg     = $extor->body;
	@background  = $extor->background;

	@links       = $extor->scheme( 'http' );
	
=head1 DESCRIPTION

This is a simple HTML link extractor designed for the person who does
not want to deal with the intricacies of C<HTML::Parser> or the
de-referencing needed to get links out of C<HTML::LinkExtor>.

You can extract all the links or some of the links (based on the HTML
tag name or attribute name). If a E<lt>BASE HREFE<gt> tag is found,
all of the relative URLs will be resolved according to that reference.

This module is simply a subclass around C<HTML::LinkExtor>, so it can
only parse what that module can handle.  Invalid HTML or XHTML may
cause problems.

If you parse multiple files, the link list grows and contains the
aggregate list of links for all of the files parsed. If you want to
reset the link list between files, use the clear_links method.

=head2 Class Methods

=over

=item $extor = HTML::SimpleLinkExtor->new()

Create the link extractor object.

=item $extor = HTML::SimpleLinkExtor->new('')
=item $extor = HTML::SimpleLinkExtor->new($base)

Create the link extractor object and resolve the relative URLs
accoridng to the supplied base URL. The supplied base URL overrides
any other base URL found in the HTML.


Create the link extractor object and do not resolve relative
links.

=cut

sub new
	{
	my $class = shift;
	my $base  = shift;

	my $self = new HTML::LinkExtor;
	bless $self, $class;

	$self->{'_SimpleLinkExtor_base'} = $base;
	$self->_init_links;
	
	return $self;
	}

=item HTML::SimpleLinkExtor->add_tags( TAG [, TAG ] )

C<HTML::SimpleLinkExtor> keeps an internal list of HTML tags (such as
'a' and 'img') that have URLs as values. If you run into another tag
that this module doesn't handle, please send it to me and I'll add it.
Until then you can add that tag to the internal list. This affects
the entire class, including previously created objects.

=cut

sub add_tags
	{
	my $self = shift;
	my $tag  = lc shift;
	
	$AUTO_METHODS{ $tag } = 'tag';
	}

=item HTML::SimpleLinkExtor->add_attributes( ATTR [, ATTR] )

C<HTML::SimpleLinkExtor> keeps an internal list of HTML tag attributes
(such as 'href' and 'src') that have URLs as values. If you run into
another attribute that this module doesn't handle, please send it to
me and I'll add it. Until then you can add that attribute to the
internal list. This affects the entire class, including previously
created objects.

=cut

sub add_attributes
	{
	my $self = shift;
	my $attr = lc shift;
	
	$AUTO_METHODS{ $attr } = 'attribute';
	}
	
=item HTML::SimpleLinkExtor->remove_tags( TAG [, TAG ] )

Take tags out of the internal list that C<HTML::SimpleLinkExtor> uses
to extract URLs. This affects the entire class, including previously
created objects.

=cut

sub remove_tags
	{
	my $self = shift;
	my $tag  = lc shift;
	
	delete $AUTO_METHODS{ $tag };
	}

=item HTML::SimpleLinkExtor->remove_attributes( ATTR [, ATTR] )

Takes attributes out of the internal list that
C<HTML::SimpleLinkExtor> uses to extract URLs. This affects the entire
class, including previously created objects.

=cut

sub remove_attributes
	{
	my $self = shift;
	my $attr = lc shift;
	
	delete $AUTO_METHODS{ $attr };
	}

=item HTML::SimpleLinkExtor->attribute_list

Returns a list of the attributes C<HTML::SimpleLinkExtor> pays
attention to.

=cut

sub attribute_list
	{
	my $self = shift;
	
	grep { $AUTO_METHODS{ $_ } eq 'attribute' } keys %AUTO_METHODS;
	}

=item HTML::SimpleLinkExtor->tag_list

Returns a list of the tags C<HTML::SimpleLinkExtor> pays attention to.

=back

=cut

sub tag_list
	{
	my $self = shift;
	
	grep { $AUTO_METHODS{ $_ } eq 'tag' } keys %AUTO_METHODS;
	}

=head2 Object methods

=over 4

=item $extor->parse_file( $filename )

Parse the file for links. Inherited from C<HTML::Parser>.

=cut


=item $extor->parse_url( $url [, $ua] )

Fetch URL and parse its content for links.

=cut

sub parse_url
	{
	my $self = shift;
	my $url  = shift;
	
	require LWP::Simple;
	
	my $data = LWP::Simple::get( $url );
	
	return unless defined $data;
	
	$self->parse( $data );
	}

=item $extor->parse( $data )

Parse the HTML in C<$data>. Inherited from C<HTML::Parser>.

=item $extor->clear_links

Clear the link list. This way, you can use the same parser for
another file.

=cut

sub clear_links { $_[0]->_init_links }

=item $extor->links

Return a list of the links.

=cut

sub links
	{
	my $self = shift;

	return map { $$_[2] } $self->_link_refs;
	}

=item $extor->img

Return a list of the links from all the SRC attributes of the
IMG.

=cut

=item $extor->frame

Return a list of all the links from all the SRC attributes of
the FRAME.

=cut

sub frames { ( $_[0]->frame, $_[0]->iframe ) }

=item $extor->iframe

Return a list of all the links from all the SRC attributes of
the IFRAME.

=item $extor->frames

Returns the combined list from frame and iframe.

=item $extor->src

Return a list of the links from all the SRC attributes of any
tag.

=item $extor->a

Return a list of the links from all the HREF attributes of the
A tags.

=item $extor->area

Return a list of the links from all the HREF attributes of the
AREA tags.

=item $extor->base

Return a list of the links from all the HREF attributes of the
BASE tags.  There should only be one.

=item $extor->href

Return a list of the links from all the HREF attributes of any
tag.

=item $extor->body, $extor->background

Return the link from the BODY tag's BACKGROUND attribute.

=item $extor->script

Return the link from the SCRIPT tag's SRC attribute

=item $extor->schemes( SCHEME, [ SCHEME, ... ] )

Return the links that use any of SCHEME. These must be absolute URLs (which
might include those converted to absolute URLs by specifying a
base). SCHEME is case-insensitive. You can specify more than one
scheme.

In list context it returns the links. In scalar context it returns
the count of the matching links.

=cut

sub schemes
	{
	my( $self, @schemes ) = @_;
	
	my %schemes;
	
	@schemes{@schemes} = lc @schemes;
	
	my @links = 
		grep { 
			my $scheme = eval { lc URI->new( $_ )->scheme };  
			exists $schemes{ $scheme };
			}
		map { $$_[2] } 
			$self->_link_refs;
			
	wantarray ? @links : scalar @links;
	}

=item $extor->absolute_links

Returns the absolute URLs (which might include those converted to
absolute URLs by specifying a base).

In list context it returns the links. In scalar context it returns
the count of the matching links.

=cut	

sub absolute_links
	{
	my $self = shift;
		
	my @links = 
		grep { 
			my $scheme = eval { lc URI->new( $_ )->scheme };  
			length $scheme;
			}
		map { $$_[2] } 
			$self->_link_refs;
	
	wantarray ? @links : scalar @links;
	}
	
=item $extor->relative_links

Returns the relatives URLs (which might exclude those converted to
absolute URLs by specifying a base or having a base in the document).

In list context it returns the links. In scalar context it returns
the count of the matching links.


=cut 

sub relative_links
	{
	my $self = shift;
		
	my @links = 
		grep { 
			my $scheme = eval { URI->new( $_ )->scheme }; 
			! defined $scheme;
			}
		map { $$_[2] } 
			$self->_link_refs;
	
	wantarray ? @links : scalar @links;
	}

=back

=head1 TO DO

This module doesn't handle all of the HTML tags that might
have links.  If someone wants those, I'll add them, or you
can edit %AUTO_METHODS in the source.

=head1 CREDITS

Will Crain who identified a problem with IMG links that had
a USEMAP attribute.

=head1 SOURCE AVAILABILITY

This source is part of a SourceForge project which always has the
latest sources in SVN, as well as all of the previous releases.

	http://sourceforge.net/projects/brian-d-foy/

If, for some reason, I disappear from the world, one of the other
members of the project can shepherd this module appropriately.

=head1 AUTHORS

brian d foy, C<< <bdfoy@cpan.org> >>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2004-2007 brian d foy.  All rights reserved.

This program is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

1;

__END__

Makefile.PL
# $Id: Makefile.PL 2426 2007-11-24 10:43:17Z comdog $
use ExtUtils::MakeMaker;

require 5.006;

eval "use Test::Manifest 1.14";   
   
WriteMakefile
	(
    'NAME'         => 'HTML::SimpleLinkExtor',
	'ABSTRACT'     => 'Extract links from HTML',
    'VERSION_FROM' => 'lib/SimpleLinkExtor.pm',
	'LICENSE'      => 'perl',
	'AUTHOR'       => 'brian d foy <bdfoy@cpan.org>',
	
 	'PM'           => {
		'lib/SimpleLinkExtor.pm'=> '$(INST_LIBDIR)/SimpleLinkExtor.pm',
		},

   'PREREQ_PM'     => { 
    	'HTML::LinkExtor' => '1.28',
		'URI'             => '1.09',
		'LWP::Simple'     => '0',
		},
		
	'MAN3PODS'     => {
		'lib/SimpleLinkExtor.pm' => '$(INST_MAN3DIR)/HTML::SimpleLinkExtor.3',
		},

	'MAN1PODS'     => {
		'scripts/linktractor' => '$(INST_MAN1DIR)/linktractor.1',
		},

	'EXE_FILES'    => [ 'scripts/linktractor' ],
	
	'dist'         => {
		'COMPRESS'        => 'gzip -9f',
		'SUFFIX'          => 'gz',
		},

    clean  => { FILES    => q|HTML-SimpleLinkExtor-*| },
	);
